File: crates/domain/src/shared/expand_events.rs
use std::collections::HashMap;

use chrono::{DateTime, Utc};
use tracing::warn;

use crate::{Calendar, CalendarEvent, EventInstance, TimeSpan, ID};

/// Generate a map of recurring_event_id to original_start_time
/// This is used to remove exceptions from the expanded events
/// The key is the recurring_event_id (as string) and the value is a vector of original_start_time
pub fn generate_map_exceptions_start_times(
    events: &Vec<CalendarEvent>,
) -> HashMap<ID, Vec<DateTime<Utc>>> {
    let mut map_recurring_event_id_to_exceptions = HashMap::new();
    for event in events {
        if let Some(recurring_event_id) = &event.recurring_event_id {
            if let Some(original_start_time) = &event.original_start_time {
                map_recurring_event_id_to_exceptions
                    .entry(recurring_event_id.clone())
                    .or_insert_with(Vec::new)
                    .push(*original_start_time);
            } else {
                warn!(
                    "Event with id: {} has a recurring_event_id but no original_start_time",
                    event.id
                );
            }
        }
    }

    map_recurring_event_id_to_exceptions
}

/// Expand an event received
/// This function will expand the event received and return a vector of EventInstance
/// This function will also remove exceptions from the expanded events
pub fn expand_event_and_remove_exceptions(
    calendar: &Calendar,
    event: &CalendarEvent,
    exceptions: &[DateTime<Utc>],
    timespan: &TimeSpan,
) -> anyhow::Result<Vec<EventInstance>> {
    let expanded_events = event.expand(Some(timespan), &calendar.settings)?;

    // If we have exceptions, remove them from the expanded events
    let expanded_events = if !exceptions.is_empty() {
        event.remove_changed_instances(expanded_events, exceptions)
    } else {
        expanded_events
    };

    Ok(expanded_events)
}

/// Expand all events received
/// This function will expand all events received and return a vector of EventInstance
/// This function will also remove exceptions from the expanded events
pub fn expand_all_events_and_remove_exceptions(
    calendars: &HashMap<String, &Calendar>,
    events: &Vec<CalendarEvent>,
    timespan: &TimeSpan,
) -> anyhow::Result<Vec<EventInstance>> {
    let map_recurring_event_id_to_exceptions = generate_map_exceptions_start_times(events);

    let mut all_expanded_events = Vec::new();
    // For each event, expand it and add the instances to the all_expanded_events
    for event in events {
        let calendar = calendars
            .get(&event.calendar_id.to_string())
            .ok_or_else(|| anyhow::anyhow!("Calendar with id: {} not found", event.calendar_id))?;
        let expanded_events = event.expand(Some(timespan), &calendar.settings)?;

        // Get the exceptions for this event
        let exceptions = map_recurring_event_id_to_exceptions.get(&event.id);

        // If we have exceptions, remove them from the expanded events
        let expanded_events = if let Some(exceptions) = exceptions {
            event.remove_changed_instances(expanded_events, exceptions)
        } else {
            expanded_events
        };

        all_expanded_events.extend(expanded_events);
    }

    Ok(all_expanded_events)
}

File: dotnet/src/Domain/Shared/ExpandEvents.cs
using Nittei.Domain;

namespace Nittei.Domain.Shared;

/// <summary>
/// Event expansion utility functions
/// </summary>
public static class ExpandEvents
{
  /// <summary>
  /// Generate a map of recurring_event_id to original_start_times (vector)
  /// This is used to remove exceptions from the expanded events
  /// The key is the recurring_event_id (as string) and the value is a vector of original_start_time
  /// </summary>
  public static Dictionary<Id, List<DateTime>> GenerateMapExceptionsOriginalStartTimes(List<CalendarEvent> events)
  {
    var mapRecurringEventIdToExceptions = new Dictionary<Id, List<DateTime>>();

    foreach (var calendarEvent in events)
    {
      if (calendarEvent.RecurringEventId.HasValue && calendarEvent.OriginalStartTime.HasValue)
      {
        var recurringEventId = calendarEvent.RecurringEventId.Value;
        var originalStartTime = calendarEvent.OriginalStartTime.Value;

        if (!mapRecurringEventIdToExceptions.ContainsKey(recurringEventId))
        {
          mapRecurringEventIdToExceptions[recurringEventId] = new List<DateTime>();
        }

        mapRecurringEventIdToExceptions[recurringEventId].Add(originalStartTime);
      }
      else if (calendarEvent.RecurringEventId.HasValue)
      {
        // Log warning: Event has recurring_event_id but no original_start_time
        Console.WriteLine($"Warning: Event with id: {calendarEvent.Id} has a recurring_event_id but no original_start_time");
      }
    }

    return mapRecurringEventIdToExceptions;
  }

  /// <summary>
  /// Expand an event received
  /// This function will expand the event received and return a vector of EventInstance
  /// This function will also remove exceptions from the expanded events
  /// </summary>
  public static List<EventInstance> ExpandEventAndRemoveExceptions(
      Calendar calendar,
      CalendarEvent calendarEvent,
      List<DateTime> exceptions,
      TimeSpan timeSpan)
  {
    // This is a simplified implementation
    // In a real implementation, you would need to implement the actual expansion logic
    var expandedEvents = new List<EventInstance>();

    // Placeholder for actual expansion logic
    // expandedEvents = calendarEvent.Expand(timeSpan, calendar.Settings);

    // If we have exceptions, remove them from the expanded events
    if (exceptions.Any())
    {
      expandedEvents = RemoveChangedInstances(expandedEvents, exceptions);
    }

    return expandedEvents;
  }

  /// <summary>
  /// Expand all events received
  /// This function will expand all events received and return a vector of EventInstance
  /// This function will also remove exceptions from the expanded events
  /// </summary>
  public static List<EventInstance> ExpandAllEventsAndRemoveExceptions(
      Dictionary<string, Calendar> calendars,
      List<CalendarEvent> events,
      TimeSpan timeSpan)
  {
    var mapRecurringEventIdToExceptions = GenerateMapExceptionsOriginalStartTimes(events);
    var allExpandedEvents = new List<EventInstance>();

    // For each event, expand it and add the instances to the all_expanded_events
    foreach (var calendarEvent in events)
    {
      if (!calendars.TryGetValue(calendarEvent.CalendarId.ToString(), out var calendar))
      {
        throw new ArgumentException($"Calendar with id: {calendarEvent.CalendarId} not found");
      }

      var exceptions = mapRecurringEventIdToExceptions
          .GetValueOrDefault(calendarEvent.Id, new List<DateTime>());

      var expandedEvents = ExpandEventAndRemoveExceptions(calendar, calendarEvent, exceptions, timeSpan);
      allExpandedEvents.AddRange(expandedEvents);
    }

    return allExpandedEvents;
  }

  /// <summary>
  /// Remove changed instances from expanded events
  /// </summary>
  private static List<EventInstance> RemoveChangedInstances(List<EventInstance> instances, List<DateTime> exceptions)
  {
    return instances.Where(instance => !exceptions.Contains(instance.StartTime)).ToList();
  }
}